### PR TITLE
[rh:requests] Patch support for `requests` 2.31+, 2.32.2+

### PR DESCRIPTION
**Pull request by Grub4K**
**IMPORTANT**: PRs without the template will be CLOSED
### Description of your *pull request* and other information

Revert [`6e36d17f`](<https://github.com/yt-dlp/yt-dlp/commit/6e36d17f404556f0e3a43f441c477a71a91877d9>) and add support for `requests` 2.32.2+ (see: [`psf/requests#6710`](<https://github.com/psf/requests/pull/6710/files>))

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Core bug fix/improvement

</details>

### CODE REVIEW

File: yt_dlp/networking/_requests.py [hunk 1]
```diff
@@ -180,14 +182,27 @@ def proxy_manager_for(self, proxy, **proxy_kwargs):
...
-        # Lean on our SSLContext for cert verification
         return self.get_connection(request.url, proxies)
 
+    # requests 2.32.2+: Reimplementation without `_urllib3_request_context`
+    def get_connection_with_tls_context(self, request, verify, proxies=None, cert=None):
+        if proxy := requests.utils.select_proxy(request.url, proxies):
+            proxy = requests.utils.prepend_scheme_if_needed(proxy, 'http')
+            if not urllib3.util.parse_url(proxy).host:
+                raise requests.exceptions.InvalidProxyURL(
+                    'Please check proxy URL. It is malformed and could be missing the host.')
```
Comment (coletdjnz) [inline yt_dlp/networking/_requests.py:200]: Technically we already validate all this already, so only need to use our select_proxy

Up to you though

**Review by coletdjnz — COMMENTED** [review]
Lgtm otherwise

File: yt_dlp/networking/_requests.py (hunk 1 above)
Comment (Grub4K) [reply yt_dlp/networking/_requests.py:200]:
```py
    def get_connection_with_tls_context(self, request, verify, proxies=None, cert=None):
        manager = self.poolmanager
        if proxy := select_proxy(request.url, proxies):
            manager = self.proxy_manager_for(proxy)

        return manager.connection_from_url(request.url)
```
this passes all the tests; do we have to normalize the url though?

Comment (coletdjnz) [reply yt_dlp/networking/_requests.py:200]: The request handler validation requires the proxy URL has a scheme

Comment (bashonly) [reply yt_dlp/networking/_requests.py:200]: do we need to normalize it with `urllib3.util.parse_url(request.url).url` though?

Comment (coletdjnz) [reply yt_dlp/networking/_requests.py:200]: Shouldn't need to but if requests does that ig it doesn't hurt

File: yt_dlp/networking/_requests.py [hunk 2]
```diff
@@ -180,14 +181,23 @@ def proxy_manager_for(self, proxy, **proxy_kwargs):
...
     def _get_connection(self, request, *_, proxies=None, **__):
-        # Lean on our SSLContext for cert verification
         return self.get_connection(request.url, proxies)
 
+    # requests 2.32.2+: Reimplementation without `_urllib3_request_context`
+    def get_connection_with_tls_context(self, request, verify, proxies=None, cert=None):
+        if proxy := select_proxy(request.url, proxies):
+            return self.proxy_manager_for(proxy).connection_from_url(request.url)
+
+        normalized_url = urllib3.util.parse_url(request.url).url
```
Comment (coletdjnz) [inline yt_dlp/networking/_requests.py:198]: Should this also be used for the proxy manager connection from URL above?

**Comment by bashonly** [comment]
- https://github.com/psf/requests/issues/6707
- https://github.com/psf/requests/issues/6715

**Pull request merged**